File: daily_challenge_leetcode/474.py
class Solution:
    def findMaxForm(self, strs: List[str], m: int, n: int) -> int:
        count = []
        for s in strs:
            countZeros = s.count("0")
            countOnes = s.count("1")
            count.append((countZeros, countOnes))
        
        # Initialize 3D memoization table: t[m][n][index]
        self.t = [[[ -1 for _ in range(len(count) + 1)] for _ in range(n + 1)] for _ in range(m + 1)]
        
        return self.solve(count, m, n, 0)
    
    def solve(self, count: List[Tuple[int, int]], m: int, n: int, index: int) -> int:
        if index >= len(count) or (m == 0 and n == 0):
            return 0
        if self.t[m][n][index] != -1:
            return self.t[m][n][index]
        
        include = 0
        if count[index][0] <= m and count[index][1] <= n:
            include = 1 + self.solve(
                count, m - count[index][0], n - count[index][1], index + 1
            )
        exclude = self.solve(count, m, n, index + 1)
        
        self.t[m][n][index] = max(include, exclude)
        return self.t[m][n][index]

File: daily_challenge_leetcode/474.cpp

// Why Greedy fails?(by sorting)
// 0/1 knap-snack will work

//Approach-1 (Recursion and Memoization)
// T.C : O(L * m * n)
//       where L = number of strings, m = max number of 0's allowed, n = max number of 1's allowed
//       For each index (L), we compute at most m*n states once due to memoization.
// S.C : O(L * m * n) for the DP cache + O(L) recursion stack

class Solution {
public:
    int t[101][101][601];
    int solve(vector<pair<int, int>>& count, int m, int n, int index) {
        if(index >= count.size() || (m == 0 && n == 0))
            return 0;
        if(t[m][n][index] != -1)
            return t[m][n][index];
        
        int include = 0;
        if(count[index].first <= m && count[index].second <= n) {
            include = 1 + solve(count, m-count[index].first, n-count[index].second, index+1);
        }
        
        int exclude = solve(count, m, n, index+1);
        
        return t[m][n][index] = max(include, exclude);
    }
    int findMaxForm(vector<string>& strs, int m, int n) {
        int N = strs.size();
        vector<pair<int, int>> count(N);
        int i = 0;\
        for(string str:strs) {
            int countOnes  = 0;
            int countZeros = 0;
            
            for(char ch:str) {
                if(ch == '1')
                    countOnes++;
                else
                    countZeros++;
            }
            
            count[i++] = {countZeros, countOnes};
        }
        
        memset(t, -1, sizeof(t));
        return solve(count, m, n, 0);
    }
};